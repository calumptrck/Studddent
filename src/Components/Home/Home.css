.viewMore {
  height: 80px;
  text-align: center;
}

.addTask-appear {
  opacity: 0.01;
  padding-bottom: 1px;
  margin-bottom: 15px;
}

.addTask-appear.addTask-appear-active {
  opacity: 1;
  padding-bottom: 18px;
  margin-bottom: 14px;
  transition: 700ms;
}

.addTask-enter {
  opacity: 0.01;
  width: 100%;
  height: 120px;
  background-color: #fff;
  border-radius: .75em;
  box-shadow: 0 2px 30px 0 hsla(0, 0%, 40%, 0.04);
  display: block;
  margin-bottom: 40px;
}

.addTask-enter.addTask-enter-active {
  opacity: 1;
  width: 100%;
  height: 150px;
  background-color: #fff;
  border-radius: .75em;
  box-shadow: 0 8px 30px 0 hsla(0, 0%, 40%, 0.08);
  display: block;
  margin-bottom: 40px;
}