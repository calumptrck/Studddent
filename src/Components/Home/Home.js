import React, { Component } from 'react';
import axios from 'axios';

import Header from '../Header/Header';
import Signup from '../Signup/Signup';
import ContentBox from '../ContentBox/ContentBox';
import Filters from '../Filters/Filters';
import Product from '../Product/Product';
import MoreButton from '../Buttons/MoreButton';
import AddButton from '../Buttons/AddButton';
import './Home.css';

const PATH_BASE = 'https://studddent.com/api/';

const isSearched = searchTerm => item =>
  item.name.toLowerCase().includes(searchTerm.toLowerCase());

const vote = (id, type) => {
  axios.post(`${PATH_BASE}votes/${type}`, {
    postid: id,
  }).catch((error) => console.log(error));
}

class Home extends Component {
  _isMounted = false; // Independent from state
  constructor() {
    super();
    this.state = {
      count: 7,
      searchTerm: '',
      buttons: [
        {
          id: 0,
          label: "ALL",
          active: true,
        },
        {
          id: 1,
          label: "DESIGN",
          active: false,
        },
        {
          id: 2,
          label: "DEVELOPMENT",
          active: false,
        },
        {
          id: 3,
          label: "UTILITY",
          active: false,
        },
      ],
      products: null,
      filteredProducts: null,
      votes: JSON.parse(localStorage.getItem('votes')) || { up: [], down: [] },
      error: null,
    }
    this.filterButtonClick = this.filterButtonClick.bind(this);
    this.upVote = this.upVote.bind(this);
    this.downVote = this.downVote.bind(this);
    this.fetchTasks = this.fetchTasks.bind(this);
    this.showMore = this.showMore.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
  }

  filterButtonClick(id) {
    let { products, filteredProducts } = this.state;
    const buttons = this.state.buttons
    for (let i = 0; i < buttons.length; i++) {
      if (i === id) {
        buttons[i].active = true;
        filteredProducts = products.filter((product) =>
          i === 0 || product.tags.indexOf(buttons[i].label.toLowerCase()) >= 0);
        this.setState({
          filteredProducts,
          count: 7,
        })
      } else {
        buttons[i].active = false;
      }
    }
    this.setState({
      buttons,
    })
  }

  upVote(id) {
    let { up, down } = this.state.votes;
    let products = this.state.products;
    let product = products.find(product => product._id === id)
    if (up.indexOf(id) < 0) {
      if (down.indexOf(id) >= 0) {
        down.splice(down.indexOf(id), 1);
        product.votes.down--;
        product.votes.up++;
        vote(id, 5);
      } else {
        product.votes.up++;
        vote(id, 3);
      }

      this.setState({
        products: [...products, product],
        votes: {
          up: [...up, id],
          down,
        }
      }, () => localStorage.setItem('votes', JSON.stringify(this.state.votes)))
    } else {
      up.splice(up.indexOf(id), 1);
      product.votes.up--;
      vote(id, 2);
      this.setState({
        products: [...products, product],
        votes: {
          up,
          down,
        }
      }, () => localStorage.setItem('votes', JSON.stringify(this.state.votes)))
    }
  }

  downVote(id) {
    let { up, down } = this.state.votes;
    let products = this.state.products;
    let product = products.find(product => product._id === id)
    if (down.indexOf(id) < 0) {
      if (up.indexOf(id) >= 0) {
        up.splice(up.indexOf(id), 1);
        product.votes.up--;
        product.votes.down++;
        vote(id, 0);
      } else {
        product.votes.down++;
        vote(id, 1);
      }
      this.setState({
        products: [...products, product],
        votes: {
          up,
          down: [...down, id]
        }
      }, () => localStorage.setItem('votes', JSON.stringify(this.state.votes)))
    } else {
      down.splice(down.indexOf(id), 1);
      product.votes.down--;
      vote(id, 4);
      this.setState({
        products: [...products, product],
        votes: {
          up,
          down,
        }
      }, () => localStorage.setItem('votes', JSON.stringify(this.state.votes)))
    }
  }

  showMore() {
    this.setState((prevState) => ({
      count: prevState.count + 7,
    })
    )
  }

  fetchTasks() {
    axios(`${PATH_BASE}links`)
      .then(result => this._isMounted && this.setState({ products: result.data, filteredProducts: result.data }))
      .catch(error => this._isMounted && this.setState({ error: error }));
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  componentDidMount() {
    this._isMounted = true;
    this.fetchTasks();
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {
    const { buttons, products, votes, count, filteredProducts } = this.state;
    return (
      <div>
        <Header>
          <a href="/"><h1 className="title">Studddent.</h1></a>
          <p className="desc">Resources for learning design & development, discounted for students.</p>
          <AddButton path={'/submit'}>Submit a Discount</AddButton>
        </Header>
        <Filters buttons={buttons} buttonClick={this.filterButtonClick} searchUpdate={this.onSearchChange} />
        <ContentBox>
          {filteredProducts &&
            filteredProducts.filter((post) =>
              post.accepted === true)
              .filter(isSearched(this.state.searchTerm))
              .slice(0, count).map((product) =>
                <Product
                  upVote={this.upVote}
                  downVote={this.downVote}
                  key={product._id}
                  product={product}
                  votes={votes} />
              )}
          <div className="viewMore">
            {products &&
              filteredProducts.filter(isSearched(this.state.searchTerm)).length > count &&
              <div>
              <MoreButton showMore={this.showMore}>VIEW MORE</MoreButton>
              <div className="spacer"></div></div>}
          </div>
          
          {products && <Signup />}
        </ContentBox>
      </div>
    );
  }
}

export default Home;