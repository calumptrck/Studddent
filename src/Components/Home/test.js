import React from 'react';
import ReactDOM from 'react-dom';
import Home from './Home';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme'
import { shallow, mount, render } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({ adapter: new Adapter() })

// Render & Snapshot

describe('Home', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Home />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Home />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Search Updates State', () => {
    const wrap = mount(
      <Home />
    )

    wrap.find('input').simulate('change', {
      target: { value: 'Microsoft' }
    })

    wrap.update()
    expect(wrap.state('searchTerm')).toEqual('Microsoft');
  });

});

// Backend Integration

const DELAY_MS = 1000

const sleep = (ms) => {
  return new Promise(resolve => setTimeout(resolve, ms));
}

const fetchTasks = async (url) => {
  try {
    const response = await fetch(url)
    const responseJson = await response.json()
    return responseJson
  }
  catch (e) {
    console.log(`fetchResponseJson failed:`, e)
  }
}

describe(`API Updates State`, () => {
  test(`on Home componentDidMount`, async () => {
    const wrapper = shallow(<Home />)
    await wrapper.instance().componentDidMount()
    await sleep(DELAY_MS+100)
    expect(wrapper.state('products')).not.toContain(null)
    expect(wrapper.state('error')).toEqual(null)

    // Force update to sync component with state
    wrapper.update()

  })

  test(`on vote`, async () => {
    const wrapper = mount(<Home />)
    await wrapper.instance().componentDidMount()
    await sleep(DELAY_MS+100)
    const expectedCount1 = wrapper.state('products')[0].votes.down;
    const expectedCount2 = expectedCount1+1;
    await wrapper.instance().upVote('5a9c3bfc1f50583c497f0ee9')
    await sleep(250);
    await wrapper.instance().downVote('5a9c3bfc1f50583c497f0ee9')
    await sleep(250);
    expect(wrapper.state('products')[0].votes.down).toEqual(expectedCount2)
    await wrapper.instance().upVote('5a9c3bfc1f50583c497f0ee9')
    await sleep(250);
    await wrapper.instance().upVote('5a9c3bfc1f50583c497f0ee9')
    await sleep(250);
    expect(wrapper.state('products')[0].votes.down).toEqual(expectedCount1)

  })

})