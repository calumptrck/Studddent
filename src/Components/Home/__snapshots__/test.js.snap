// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home has a valid snapshot 1`] = `
<div>
  <div
    className="header"
  >
    <div
      className="container"
    >
      <a
        href="/"
      >
        <h1
          className="title"
        >
          Studddent.
        </h1>
      </a>
      <p
        className="desc"
      >
        Best online student discounts. Resources for learning design & development.
      </p>
      <a
        href="/submit"
      >
        <button
          className="btn addDiscount grow2 shadow2"
          type="button"
        >
          Submit a Discount
        </button>
      </a>
    </div>
  </div>
  <div
    className="container filtersContainer"
  >
    <input
      aria-describedby="search"
      autoComplete="off"
      id="search"
      name="email"
      onChange={[Function]}
      placeholder="Search for products"
      type="text"
    />
    <div
      className="filters"
    >
      <button
        className="filterButton default"
        id={undefined}
        onClick={[Function]}
        type="button"
      >
        ALL
      </button>
      <button
        className="filterButton default"
        id={undefined}
        onClick={[Function]}
        type="button"
      >
        DESIGN
      </button>
      <button
        className="filterButton default"
        id={undefined}
        onClick={[Function]}
        type="button"
      >
        DEVELOPMENT
      </button>
      <button
        className="filterButton default"
        id={undefined}
        onClick={[Function]}
        type="button"
      >
        UTILITY
      </button>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="contentBox"
    >
      <div
        className="viewMore"
      />
    </div>
  </div>
</div>
`;
